driver/src/price_estimation/mod.rs

```diff
@@ -1,30 +1,151 @@
 //! Module responsible for aggregating price estimates from various sources to
 //! give good price estimates to the solver for better results.
 
-#[allow(dead_code)]
 mod kraken;
 
-use crate::models::{TokenId, TokenInfo};
+use self::kraken::KrakenClient;
+use crate::models::{Order, TokenData, TokenId, TokenInfo};
 use anyhow::Result;
-use ethcontract::H160;
-use std::collections::HashMap;
+use lazy_static::lazy_static;
+use log::warn;
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::iter;
+
+/// A type alias for token information map that is passed to the solver.
+type Tokens = BTreeMap<TokenId, Option<TokenInfo>>;
+
+/// A trait representing a price oracle that retrieves price estimates for the
+/// tokens included in the current orderbook.
+#[cfg_attr(test, mockall::automock)]
+pub trait PriceEstimating {
+    fn get_token_prices(&self, orders: &[Order]) -> Tokens;
+}
+
+pub struct PriceOracle {
+    /// The token data supplied by the environment. This ensures that only
```

Comment:
I would prefer not assuming that a token is whitelisted for the external price feed if the token info is provided. Simply for the reason that we might want to provide data for tokens, which are legitimate, but still there is not price feed available. And easy example would be SNX, as there is barely any centralized exchange trading this token.

But we could add another field to the tokenInfo: `LegibleForAPIPriceProvision` TRUE/FALSE, and then split the list `(mut tokens_to_price, mut unpriced_token_ids)` with this boolean information

Reply:
I added a `shouldEstimatePrice` flag that, when set to `true` will enable price estimations for that token. Note that it still parses the old token data format if the option is omitted, defaulting to false.

Reply:
I implemented this as part of a separate PR #578 to keep them easier to review.